core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/SingleDirectionalParentVariableReferenceGraph.java

```diff
@@ -0,0 +1,149 @@
+            Object[] entities) {
+        variableReferenceToInstanceMap = CollectionUtils.newHashMap(sortedDeclarativeShadowVariableDescriptors.size());
+        sortedVariableUpdaterInfoList = new ArrayList<>(sortedDeclarativeShadowVariableDescriptors.size());
+        monitoredSourceVariableSet = new HashSet<>();
+        changedSet = Collections.newSetFromMap(CollectionUtils.newIdentityHashMap(entities.length));
+
+        this.successorFunction = successorFunction;
+        this.comparator = comparator;
+        this.changedVariableNotifier = changedVariableNotifier;
+        this.loopedDescriptor =
+                sortedDeclarativeShadowVariableDescriptors.get(0).getEntityDescriptor().getShadowVariableLoopedDescriptor();
+
+        for (var variableDescriptor : sortedDeclarativeShadowVariableDescriptors) {
+            var variableMetaModel = variableDescriptor.getVariableMetaModel();
+            var objectMap = CollectionUtils.newIdentityHashMap(entities.length);
+            var variableUpdaterInfo = new VariableUpdaterInfo<>(
+                    variableMetaModel,
+                    variableDescriptor,
+                    loopedDescriptor,
+                    variableDescriptor.getMemberAccessor(),
+                    variableDescriptor.getCalculator()::executeGetter);
+            sortedVariableUpdaterInfoList.add(variableUpdaterInfo);
+            variableReferenceToInstanceMap.put(variableMetaModel, CollectionUtils.newIdentityHashMap(entities.length));
+
+            for (var source : variableDescriptor.getSources()) {
+                for (var sourceReference : source.variableSourceReferences()) {
+                    monitoredSourceVariableSet.add(sourceReference.variableMetaModel());
+                }
+            }
+
+            for (var entity : entities) {
+                // No graph, so no graph id
+                if (variableDescriptor.getEntityDescriptor().getEntityClass().isInstance(entity)) {
+                    objectMap.put(entity, new EntityVariablePair<>(entity, variableUpdaterInfo, -1));
+                    changedSet.add(entity);
+                }
+            }
+        }
+    }
+
+    @Override
+    public @Nullable EntityVariablePair<Solution_> lookupOrNull(VariableMetaModel<?, ?, ?> variableId, Object entity) {
+        return variableReferenceToInstanceMap.get(variableId).get(entity);
+    }
+
+    @Override
+    public void addEdge(@NonNull EntityVariablePair<Solution_> from, @NonNull EntityVariablePair<Solution_> to) {
+        markChanged(to);
+    }
+
+    @Override
+    public void removeEdge(@NonNull EntityVariablePair<Solution_> from, @NonNull EntityVariablePair<Solution_> to) {
+        markChanged(to);
+    }
+
+    @Override
+    public void markChanged(@NonNull EntityVariablePair<Solution_> node) {
+        changedSet.add(node.entity());
+    }
+
+    @Override
+    public void updateChanged() {
+        var changedSorted = new TreeSet<>(comparator);
```

Comment:
Can't we add it to the `changedSorted` immediately?
Seems like we just waste cycles, adding it to a different collection and only then creating another collection from it.